### PR TITLE
[WIP] [Cache] Allow to chain adapters

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | todo |

Allows to chain adapters (e.g: use the `ArrayAdapter` and fallback to the `DoctrineAdapter`).
- [ ] `ArrayAdapter`
- [ ] `ProxyAdapter`

### CODE REVIEW

**Comment by DavidBadura** [comment]
I think the decorator pattern is the wrong way. All adapters must be able to handle it.
A chain class is as much better or not? 

Like this: https://github.com/php-cache/chain-adapter/blob/master/src/CachePoolChain.php

**Comment by dunglas** [comment]
The chain suits probably better to this case, indeed.

**Comment by dunglas** [comment]
I was thinking that all adapters was extending the abstract class but it's not the case (and it's why this PR is marked as WIP).

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 1]
```diff
@@ -109,7 +129,6 @@ function ($deferred, $namespace) {
     public function getItem($key)
     {
         $id = $this->getId($key);
-
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: should be reverted

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 2]
```diff
@@ -12,23 +12,43 @@
...
 
 /**
  * @author Nicolas Grekas <p@tchwork.com>
  */
-abstract class AbstractAdapter implements CacheItemPoolInterface
+abstract class AbstractAdapter implements AdapterInterface
 {
+    /**
+     * @var string
+     */
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: I would remove these docblocks. They are pretty useless as IDEs should be able to infer the type from the initialisation in the constructor.

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 3]
```diff
@@ -169,7 +199,13 @@ public function hasItem($key)
             $this->commit();
         }
 
-        return $this->doHave($this->getId($key));
+        $got = $this->doHave($this->getId($key));
+
+        if ($got || !$this->decorated) {
```
Comment (xabbuh) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: In the other places we have this condition the other way around (checking if we did not find anything and then calling the decorated adapter). For consistency, I suggest to do the same here too.

**Comment by dunglas** [comment]
Replaced by #17556.

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php (hunk 2 above)
Comment (iltar) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: You're not always looking at the code via your IDE. Imo those doc blocks are valuable until you can do a `private AdapterInterface $decorated;` for example.

Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: I would remove them as well. They are verbose, annoying and don't add anything useful.

File: src/Symfony/Component/Cache/Adapter/AbstractAdapter.php [hunk 4]
```diff
@@ -12,23 +12,43 @@
...
+     */
     private $namespace;
+
+    /**
+     * @var AdapterInterface
+     */
+    private $decorated;
+
+    /**
+     * @var array
```
Comment (fabpot) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: Same here. Just by looking at the default value, I know this is an array.

Comment (dunglas) [inline src/Symfony/Component/Cache/Adapter/AbstractAdapter.php]: This PR has been replaced by https://github.com/symfony/symfony/pull/17556. The new one doesn't include docblocks changes anymore.

**Pull request closed without merging**